crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs

```diff
@@ -127,31 +129,164 @@ impl Extension for QueryLimitsChecker {
         let cfg = ctx
             .data::<ServiceConfig>()
             .expect("No service config provided in schema data");
-        if query.len() > cfg.limits.max_query_payload_size as usize {
-            metrics
-                .request_metrics
-                .query_payload_too_large_size
-                .observe(query.len() as f64);
-            info!(
-                query_id = %query_id,
-                session_id = %session_id,
-                error_code = code::BAD_USER_INPUT,
-                "Query payload is too large: {}",
-                query.len()
-            );
-
-            return Err(graphql_error(
-                code::BAD_USER_INPUT,
-                format!(
-                    "Query payload is too large. The maximum allowed is {} bytes",
-                    cfg.limits.max_query_payload_size
-                ),
-            ));
-        }
-
         // Document layout of the query
         let doc = next.run(ctx, query, variables).await?;
 
+        let mutation_node = doc
+            .operations
+            .iter()
+            .find(|(_, operation)| operation.node.ty == OperationType::Mutation);
+
+        let is_mutation = mutation_node.is_some();
+
+        // this is a map of type node position, (tx_bytes, signatures) bytes for all nodes of type
+        // executeTransactionBlock
+        let mut tx_bytes_sigs = BTreeMap::new();
+
+        // A few checks in case it is a mutation
+        if is_mutation {
+            // get the mutation node and the selection set
+            let mutation_selection_set =
+                mutation_node.map(|(_, operation)| &operation.node.selection_set);
+
+            // the number of bytes for the tx_bytes and signatures strings
+            // Case 1: check if the txBytes for each executeTransactionBlock node is larger than
+            // the allowed limit
+            if let Some(selection_set) = mutation_selection_set {
+                for x in selection_set.node.items.iter() {
+                    if let Selection::Field(f) = &x.node {
+                        if f.node.name.node == "executeTransactionBlock"
+                            || f.node.name.node == "dryRunTransactionBlock"
+                        {
+                            // we know this has always two arguments as per the schema
+                            let tx_bytes_arg = &f.node.arguments[0].1.node;
+                            let sigs_arg = &f.node.arguments[1].1.node;
+                            let tx_bytes_len = get_value_str_len(tx_bytes_arg, variables);
+                            let sigs_len = get_value_str_len(sigs_arg, variables);
```

Comment:
Can we perform this check on some other hook, where async graphql has already done variable resolution for us?

Reply:
I've looked into doing this in the `validation` hook, but I wasn't able to figure out how to get the resolved query. 
https://async-graphql.github.io/async-graphql/en/extensions_inner_working.html#validation

Reply:
I am not sure it's possible, but if you have a hint it'd be useful.